### ISSUE TITLE
[Bug Report] UnobservedTaskException (ObjectDisposedException) from Microsoft.Azure.Amqp in AmqpAuthenticationRefresher

### ISSUE THREAD

**Issue by ntrpnr** [Azure/azure-iot-sdk-csharp#2243]
# Context

- **OS, version, SKU and CPU architecture used:** Windows 11 Desktop x64
- **Application's .NET Target Framework :** net472
- **Device:** Desktop
- **SDK version used:** Microsoft.Azure.Devices.Client 1.39.0, Microsoft.Azure.Amqp 2.4.7

## Description of the issue
We intermittently experience this exception in Microsoft.Azure.Devices.Client:
```
System.AggregateException: A Task's exception(s) were not observed either by Waiting on the Task or accessing its Exception property. As a result, the unobserved exception was rethrown by the finalizer thread. ---> System.ObjectDisposedException: Cannot access a disposed object.
Object name: '$cbs'.
   at Microsoft.Azure.Amqp.AmqpCbsLink.BeginSendToken(ICbsTokenProvider tokenProvider, Uri namespaceAddress, String audience, String resource, String[] requiredClaims, TimeSpan timeout, AsyncCallback callback, Object state)
   at Microsoft.Azure.Amqp.AmqpCbsLink.<>c__DisplayClass4_0.<SendTokenAsync>b__0(AsyncCallback c, Object s)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl(Func`3 beginMethod, Func`2 endFunction, Action`1 endAction, Object state, TaskCreationOptions creationOptions)
   at Microsoft.Azure.Amqp.TaskHelpers.CreateTask[T](Func`3 begin, Func`2 end, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.AmqpIot.AmqpIotCbsLink.<SendTokenAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpAuthenticationRefresher.<RefreshLoopAsync>d__12.MoveNext()
   --- End of inner exception stack trace ---
---> (Inner Exception #0) System.ObjectDisposedException: Cannot access a disposed object.
Object name: '$cbs'.
   at Microsoft.Azure.Amqp.AmqpCbsLink.BeginSendToken(ICbsTokenProvider tokenProvider, Uri namespaceAddress, String audience, String resource, String[] requiredClaims, TimeSpan timeout, AsyncCallback callback, Object state)
   at Microsoft.Azure.Amqp.AmqpCbsLink.<>c__DisplayClass4_0.<SendTokenAsync>b__0(AsyncCallback c, Object s)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl(Func`3 beginMethod, Func`2 endFunction, Action`1 endAction, Object state, TaskCreationOptions creationOptions)
   at Microsoft.Azure.Amqp.TaskHelpers.CreateTask[T](Func`3 begin, Func`2 end, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.AmqpIot.AmqpIotCbsLink.<SendTokenAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpAuthenticationRefresher.<RefreshLoopAsync>d__12.MoveNext()<---
```

## Code sample exhibiting the issue
We are unable to share our code in full, and since this error is highly intermittent we are not able to produce sample code.
Our implementation is very similar to this code:
https://github.com/Azure-Samples/azure-iot-samples-csharp/blob/413415e8f6abced86b73e23faf755d5501e1bc34/iot-hub/Samples/device/DeviceReconnectionSample/DeviceReconnectionSample.cs

## Console log of the issue
Unable to produce console log due to issue being highly intermittent.

**Comment by ntrpnr** [comment]
This exception occurs every 4-5 hours, _possibly_ only on idle (when no data is being sent or received). 
As this is an `UnobservedTaskException` it does not terminate our application, but it seems to drop the connection (further research needed).

Currently we have a workaround which is something like this:

```
TaskScheduler.UnobservedTaskException += TaskScheduler_UnobservedTaskException;

private void TaskScheduler_UnobservedTaskException(object sender, UnobservedTaskExceptionEventArgs e)
{
	reInitializeDeviceClientManagerCallback();
}
```

**Comment by jamdavi** [comment]
Thanks @ntrpnr is the callstack in the issue description from the UnobservedTaskException event, or did you find it another way?

**Comment by ntrpnr** [comment]
@jamdavi Yes, basically:
private static void TaskScheduler_UnobservedTaskException(object sender, UnobservedTaskExceptionEventArgs e)
{
	Exception ex = (Exception)e.Exception;
	//...
}

**Comment by jamdavi** [comment]
@ntrpnr Thanks. I am going to run an experiment with the current release version and vNext to see if I can reproduce this issue and if this exists in the next version.

In the next version we've updated to Amqp 2.5.8 and that version has a much cleaner await/async chain.

**Comment by jamdavi** [comment]
@ntrpnt, this matches my tests as well. I was going to give it until the end of the week to be sure, but I'm glad you were able to confirm.

As a note we have this in main right now so it will go with the next release.

https://github.com/Azure/azure-iot-sdk-csharp/blob/main/iothub/device/src/Microsoft.Azure.Devices.Client.csproj#L99

**Comment by ntrpnr** [comment]
@jamdavi Just out of curiosity, were you able to reproduce the issue? If so, did it affect the connection?

**Comment by jamdavi** [comment]
It reproduced once for me but it was not a consistent experience.

**Comment by jamdavi** [comment]
This is now fixed with version 1.40.0 [https://github.com/Azure/azure-iot-sdk-csharp/releases/tag/2022_01_26].
